core/optaplanner-core-impl/src/main/java/org/optaplanner/core/impl/localsearch/decider/MultiThreadedLocalSearchDecider.java

```diff
@@ -180,18 +181,6 @@ public void decideNextStep(LocalSearchStepScope<Solution_> stepScope) {
     }
 
     private boolean forageResult(LocalSearchStepScope<Solution_> stepScope, int stepIndex) {
-        if (resultQueue.isEmpty() && operationQueue.size() <= moveThreadCount) {
-            /*
-             * When the phase starts, there might be up to _moveThreadCount_ unprocessed SetupOperations.
-             * Similarly, when the phase ends, there might be up to _moveThreadCount_ unprocessed DestroyOperations.
-             * If there is no MoveEvaluationOperation in the _operationQueue_, the _resultQueue_ will remain empty.
-             */
-            boolean operationQueueContainsNoMoveEvaluation = operationQueue.stream()
-                    .noneMatch(moveThreadOperation -> moveThreadOperation instanceof MoveEvaluationOperation);
-            if (operationQueueContainsNoMoveEvaluation) {
-                return true;
-            }
-        }
```

Comment:
This was the code with the race condition.

Reply:
The race condition exists because sometimes it may happen that both the queues are empty and there is still some move being evaluated, right?

Reply:
2 reasons:
- yes, indeed, because  both the queues are empty and there is still some move being evaluated
- because the 2 checks are synchronized (which would cause massive congestion). See Briant Goetz Concurrency in practice book. It's not just about changes between the two isEmpty() checks, but also about CPU cache staleness, etc.

Reply:
Is the second point relevant here? I think the condition is evaluated by a single thread.